vpn/speaker.go

```diff
@@ -0,0 +1,370 @@
+	serdes    *serdes[S, R, RR]
+	requests  chan *request[S, R]
+	logger    slog.Logger
+	nextMsgID uint64
+
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	sendCh       chan<- S
+	recvCh       <-chan R
+	recvLoopDone chan struct{}
+
+	mu            sync.Mutex
+	responseChans map[uint64]chan R
+}
+
+// newSpeaker creates a new protocol speaker.
+func newSpeaker[S rpcMessage, R receivableRPCMessage[RR], RR any](
+	ctx context.Context, logger slog.Logger, conn io.ReadWriteCloser,
+	me, them SpeakerRole,
+) (
+	*speaker[S, R, RR], error,
+) {
+	ctx, cancel := context.WithCancel(ctx)
+	if err := handshake(ctx, conn, logger, me, them); err != nil {
+		cancel()
+		return nil, xerrors.Errorf("handshake failed: %w", err)
+	}
+	sendCh := make(chan S)
+	recvCh := make(chan R)
+	s := &speaker[S, R, RR]{
+		serdes:        newSerdes(ctx, logger, conn, sendCh, recvCh),
+		logger:        logger,
+		requests:      make(chan *request[S, R]),
+		responseChans: make(map[uint64]chan R),
+		nextMsgID:     1,
+		ctx:           ctx,
+		cancel:        cancel,
+		sendCh:        sendCh,
+		recvCh:        recvCh,
+		recvLoopDone:  make(chan struct{}),
+	}
+	return s, nil
+}
+
+// start starts the serialzation/deserialization.  It's important this happens
+// after any assignments of the speaker to its owning Tunnel or Manager, since
+// the mutex is copied and that is not threadsafe.
+// nolint: revive
+func (s *speaker[_, _, _]) start() {
+	s.serdes.start()
+	go s.recvFromSerdes()
+}
+
+func (s *speaker[S, R, _]) recvFromSerdes() {
+	defer close(s.recvLoopDone)
+	defer close(s.requests)
+	for {
+		select {
+		case <-s.ctx.Done():
+			s.logger.Debug(s.ctx, "recvFromSerdes context done while waiting for proto", slog.Error(s.ctx.Err()))
+			return
+		case msg, ok := <-s.recvCh:
+			if !ok {
+				s.logger.Debug(s.ctx, "recvCh is closed")
+				return
+			}
+			rpc := msg.GetRpc()
+			if rpc != nil && rpc.ResponseTo != 0 {
+				// this is a unary response
+				s.tryToDeliverResponse(msg)
+				continue
+			}
+			req := &request[S, R]{
+				ctx:     s.ctx,
+				msg:     msg,
+				replyCh: s.sendCh,
+			}
+			select {
+			case <-s.ctx.Done():
+				s.logger.Debug(s.ctx, "recvFromSerdes context done while waiting for request handler", slog.Error(s.ctx.Err()))
+				return
+			case s.requests <- req:
+			}
+		}
+	}
+}
+
+// nolint: revive
+func (s *speaker[_, _, _]) Close() error {
+	s.cancel()
+	err := s.serdes.Close()
+	return err
+}
+
+// unaryRPC sends a request/response style RPC over the protocol, waits for the response, then
+// returns the response
+func (s *speaker[S, R, _]) unaryRPC(ctx context.Context, req S) (resp R, err error) {
+	rpc := req.EnsureRPC()
+	msgID, respCh := s.newRPC()
+	rpc.MsgId = msgID
+	logger := s.logger.With(slog.F("msg_id", msgID))
+	select {
+	case <-ctx.Done():
+		return resp, ctx.Err()
+	case <-s.ctx.Done():
+		return resp, xerrors.Errorf("vpn protocol closed: %w", s.ctx.Err())
+	case s.sendCh <- req:
```

Comment:
Until one of the contexts exits. I guess I could also read `recvLoopDone` here, since `serdes` exiting on the send loop closes the recv side too, and that'll bubble up to the speaker.